dipy/nn/evac.py

```diff
@@ -0,0 +1,455 @@
+#!/usr/bin/python
+"""
+Class and helper functions for fitting the EVAC+ model.
+"""
+
+
+from packaging.version import Version
+import logging
+import numpy as np
+from scipy.ndimage import affine_transform
+from dipy.data import get_fnames
+from dipy.testing.decorators import doctest_skip_parser
+from dipy.utils.optpkg import optional_package
+from dipy.io.image import load_nifti
+from dipy.align.reslice import reslice
+from dipy.nn.utils import normalize
+
+tf, have_tf, _ = optional_package('tensorflow')
+if have_tf:
+    from tensorflow.keras.models import Model
+    from tensorflow.keras.layers import Softmax, Conv3DTranspose
+    from tensorflow.keras.layers import Conv3D, LayerNormalization, LeakyReLU
+    from tensorflow.keras.layers import Concatenate, Layer, Dropout, Add
+    if Version(tf.__version__) < Version('2.0.0'):
+        raise ImportError('Please upgrade to TensorFlow 2+')
+else:
+    class Model:
+        pass
+
+    class Layer:
+        pass
+    logging.warning('This model requires Tensorflow.\
+                    Please install these packages using \
+                    pip. If using mac, please refer to this \
+                    link for installation. \
+                    https://github.com/apple/tensorflow_macos')
+
+logging.basicConfig()
+logger = logging.getLogger('EVAC+')
+
+
+def set_logger_level(log_level):
+    """ Change the logger of the EVAC+ to one on the following:
+    DEBUG, INFO, WARNING, CRITICAL, ERROR
+
+    Parameters
+    ----------
+    log_level : str
+        Log level for the EVAC+ only
+    """
+    logger.setLevel(level=log_level)
+
+def transform_img(image, affine):
+    r"""
+    Function to reshape image as an input to the model
+
+    Parameters
+    ----------
+    image : np.ndarray
+    affine : np.ndarray
+    voxsize : tuple
+
+    Returns
+    -------
+    transformed_img : np.ndarray
+    """
+    affine2 = affine.copy()
+    affine2[:3, 3] += np.array([128, 128, 128])
+    inv_affine = np.linalg.inv(affine2)
+    transformed_img = affine_transform(image, inv_affine, output_shape=(256, 256, 256))
+    transformed_img, _ = reslice(transformed_img, np.eye(4), (1, 1, 1), (2, 2, 2))
+    return transformed_img, affine2
+    
+def recover_img(image, affine, ori_shape):
+    r"""
+    Function to recover image back to its original shape
+
+    Parameters
+    ----------
+    image : np.ndarray
+    affine : np.ndarray
+    voxsize : tuple
+    ori_shape : tuple
+
+    Returns
+    -------
+    recovered_img : np.ndarray
+    """
+    new_image, _ = reslice(image, np.eye(4), (2, 2, 2), (1, 1, 1))
+    recovered_img = affine_transform(new_image, affine, output_shape=ori_shape)
+    return recovered_img
```

Comment:
why all this function are not in `nn.utils`?    I mean `prepare_img`, `recover_img`, `transform_img`, `set_logger_level` ?

Reply:
oh,I see , many fixed size...  maybe they can be set as default arguments